### ISSUE TITLE
HAProxy 2.9.3: Current worker (..) exited with code 139 (Segmentation fault)

### ISSUE THREAD

**Issue by joachimbuechse** [haproxy/haproxy#2417]
### Detailed Description of the Problem

During performance evaluation HAProxy restarts indicating segmentation fault in one worker after less than 1M requests.

Problem is reproducible with

h2load --threads=16 --clients=1000 --requests=5000000  --log-file=/var/log/h2load_`date +"%d-%m-%Y_%H-%M-%S"`.log https://[snip]/epg/v009/programs/now

### Expected Behavior

HAProxy runs stable

### Steps to Reproduce the Behavior

1. start HAProxy (see config attached)
2. start load test (load generator is on a 32 core amazon instance generating ~4 Gbps of incoming traffic, ~8000 req/s) h2load --threads=16 --clients=1000 --requests=5000000  --log-file=/var/log/h2load_`date +"%d-%m-%Y_%H-%M-%S"`.log https://[snip]/epg/v009/programs/now

### What is your configuration?

```haproxy
global
	log /dev/log	local0
	log /dev/log	local1 notice
	chroot /var/lib/haproxy
	stats socket /run/haproxy/admin.sock mode 660 level admin
	stats timeout 30s
	user haproxy
	group haproxy
	daemon
	maxconn 2000

	# Load CORS handling module
	lua-load /etc/haproxy/cors_tv.lua

	# Default SSL material locations
	ca-base /etc/ssl/certs
	crt-base /etc/ssl/private

	# See: https://ssl-config.mozilla.org/#server=haproxy&server-version=2.0.3&config=intermediate
        ssl-default-bind-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384
        ssl-default-bind-ciphersuites TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256
        ssl-default-bind-options ssl-min-ver TLSv1.2 no-tls-tickets

defaults
	log	global
	mode	http
	option	httplog
	option	dontlognull
        timeout connect 5000
        timeout client  50000
        timeout server  50000
	errorfile 400 /etc/haproxy/errors/400.http
	errorfile 403 /etc/haproxy/errors/403.http
	errorfile 408 /etc/haproxy/errors/408.http
	errorfile 500 /etc/haproxy/errors/500.http
	errorfile 502 /etc/haproxy/errors/502.http
	errorfile 503 /etc/haproxy/errors/503.http
	errorfile 504 /etc/haproxy/errors/504.http

frontend stats
	mode http
	bind 10.15.28.32:80
	stats enable
	stats uri /stats
	stats admin if LOCALHOST                                                                                                         

frontend fe-epgMS
	maxconn 2000
	bind [snip]:80
	bind [snip]:443 ssl crt /etc/haproxy/certs/[snip].bundle.pem alpn h2,http/1.1

	# define services
	acl is_epg path_beg /epg/

	# remove headers we dont want to forward
	http-request del-header X-Forwarded-For
	# TODO improve: we want to remove the cache busters
	http-request set-query %[query,regsub(client=web\d+,client=web,g)]

	# log special headers
	capture request header x-quickline-userid len 20
	capture request header x-quickline-profileid len 36
	capture request header x-quickline-deviceid len 36
	capture request header x-quickline-function len 60

	# handle preflight requests
	http-request lua.cors "GET,PUT,POST,PATCH,DELETE" "[snip]" "Authorization,cache-control,pragma,x-quickline-deviceid,x-quickline-profileid,x-quickline-userid[snip]"

	use_backend be-epgBulkServersCanaryPreprod if is_epg { req.hdr(host) -i temp-epg.qltv.quickline.ch }
	use_backend be-deny


cache cache-epgBulkCanaryPreprod
	total-max-size 1024
	max-object-size 4000000
	max-age 120
	process-vary on
	max-secondary-entries 12

backend be-epgBulkServersCanaryPreprod
	option httpchk
	http-check connect port 8081
	http-check send meth GET uri /ha-check
	http-check expect status 200
	http-check expect string alive

	balance roundrobin

	option forwardfor

	option http-buffer-request
	option http-keep-alive
	http-reuse always

	# haproxy 2.9 only accepts Vary if the header list contains Accept-Encoding..
	http-response del-header Vary
	http-response add-header Vary "Origin, Accept-Encoding"
	http-request cache-use cache-epgBulkCanaryPreprod
	http-response cache-store cache-epgBulkCanaryPreprod

	# limit number of concurrent request to each server to 20
	server clpreprod-epgBulk01 [snip]:8081 check maxconn 20

backend be-deny
	http-request deny
```

### Output of `haproxy -vv`

```plain
# uname -a
Linux slb-epg21 6.1.0-17-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.69-1 (2023-12-30) x86_64 GNU/Linux

# haproxy -vv
HAProxy version 2.9.3-1~bpo12+1 2024/01/18 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2025.
Known bugs: http://www.haproxy.org/bugs/bugs-2.9.3.html
Running on: Linux 6.1.0-17-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.69-1 (2023-12-30) x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_SYSTEMD=1 USE_OT=1 USE_QUIC=1 USE_PROMEX=1 USE_PCRE2=1 USE_PCRE2_JIT=1 USE_QUIC_OPENSSL_COMPAT=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE -LIBATOMIC +LIBCRYPT +LINUX_CAP +LINUX_SPLICE +LINUX_TPROXY +LUA +MATH -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_AWSLC -OPENSSL_WOLFSSL +OT -PCRE +PCRE2 +PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX -PTHREAD_EMULATION +QUIC +QUIC_OPENSSL_COMPAT +RT +SHM_OPEN +SLZ +SSL -STATIC_PCRE -STATIC_PCRE2 +SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=2).
Built with OpenSSL version : OpenSSL 3.0.11 19 Sep 2023
Running on OpenSSL version : OpenSSL 3.0.11 19 Sep 2023
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Built with Lua version : Lua 5.3.6
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with OpenTracing support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.42 2022-12-11
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 12.2.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
       quic : mode=HTTP  side=FE     mux=QUIC  flags=HTX|NO_UPG|FRAMED
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[BWLIM] bwlim-in
	[BWLIM] bwlim-out
	[CACHE] cache
	[COMP] compression
	[FCGI] fcgi-app
	[  OT] opentracing
	[SPOE] spoe
	[TRACE] trace
```

### Last Outputs and Backtraces

```plain
# egrep -A 1 'haproxy.+segfault' /var/log/syslog
2024-01-18T11:23:05.600716+01:00 slb-epg21 kernel: [651297.405105] haproxy[12925]: segfault at 8 ip 000055ca1789dc10 sp 00007ffccf8c2bd0 error 4 in haproxy[55ca176dd000+267000] likely on CPU 0 (core 0, socket 0)
2024-01-18T11:23:05.600717+01:00 slb-epg21 kernel: [651297.405124] Code: 00 00 48 8b 40 0f a8 01 74 20 66 83 78 1f 00 78 f1 48 85 c9 74 47 83 69 68 01 e8 eb 39 0a 00 c7 43 3c 00 00 00 00 5b c3 66 90 <48> 8b 50 08 48 83 fa 01 76 dd 66 83 78 20 00 79 d6 f6 c2 01 74 18
--
2024-01-18T15:16:00.198674+01:00 slb-epg21 kernel: [665270.716712] haproxy[13795]: segfault at 8 ip 000055ed5d091c10 sp 00007f065eff32a0 error 4 in haproxy[55ed5ced1000+267000] likely on CPU 0 (core 0, socket 0)
2024-01-18T15:16:00.198689+01:00 slb-epg21 kernel: [665270.716727] Code: 00 00 48 8b 40 0f a8 01 74 20 66 83 78 1f 00 78 f1 48 85 c9 74 47 83 69 68 01 e8 eb 39 0a 00 c7 43 3c 00 00 00 00 5b c3 66 90 <48> 8b 50 08 48 83 fa 01 76 dd 66 83 78 20 00 79 d6 f6 c2 01 74 18
--
2024-01-19T09:56:45.050741+01:00 slb-epg21 kernel: [732509.377131] haproxy[14262]: segfault at 8 ip 00005604dcc59c10 sp 00007f549d1f32a0 error 4 in haproxy[5604dca99000+267000] likely on CPU 1 (core 0, socket 2)
2024-01-19T09:56:45.050776+01:00 slb-epg21 kernel: [732509.377195] Code: 00 00 48 8b 40 0f a8 01 74 20 66 83 78 1f 00 78 f1 48 85 c9 74 47 83 69 68 01 e8 eb 39 0a 00 c7 43 3c 00 00 00 00 5b c3 66 90 <48> 8b 50 08 48 83 fa 01 76 dd 66 83 78 20 00 79 d6 f6 c2 01 74 18
--
2024-01-19T11:38:32.894682+01:00 slb-epg21 kernel: [738616.657885] haproxy[15299]: segfault at 8 ip 00005618a8563c10 sp 00007ffec4c02c50 error 4 in haproxy[5618a83a3000+267000] likely on CPU 1 (core 0, socket 2)
2024-01-19T11:38:32.894700+01:00 slb-epg21 kernel: [738616.658082] Code: 00 00 48 8b 40 0f a8 01 74 20 66 83 78 1f 00 78 f1 48 85 c9 74 47 83 69 68 01 e8 eb 39 0a 00 c7 43 3c 00 00 00 00 5b c3 66 90 <48> 8b 50 08 48 83 fa 01 76 dd 66 83 78 20 00 79 d6 f6 c2 01 74 18
--
2024-01-19T11:57:00.124078+01:00 slb-epg21 kernel: [739723.785353] haproxy[15623]: segfault at 8 ip 0000555d27456880 sp 00007ffce0a4ca90 error 4 in haproxy[555d27295000+267000] likely on CPU 0 (core 0, socket 0)
2024-01-19T11:57:00.124090+01:00 slb-epg21 kernel: [739723.785378] Code: 00 00 48 8b 40 0f a8 01 74 20 66 83 78 1f 00 78 f1 48 85 c9 74 47 83 69 68 01 e8 cb 39 0a 00 c7 43 3c 00 00 00 00 5b c3 66 90 <48> 8b 50 08 48 83 fa 01 76 dd 66 83 78 20 00 79 d6 f6 c2 01 74 18
--
2024-01-19T14:35:00.126691+01:00 slb-epg21 kernel: [749202.916188] haproxy[15738]: segfault at 8 ip 000055b1445b7880 sp 00007fff2449bf20 error 4 in haproxy[55b1443f6000+267000] likely on CPU 1 (core 0, socket 2)
2024-01-19T14:35:00.126699+01:00 slb-epg21 kernel: [749202.916207] Code: 00 00 48 8b 40 0f a8 01 74 20 66 83 78 1f 00 78 f1 48 85 c9 74 47 83 69 68 01 e8 cb 39 0a 00 c7 43 3c 00 00 00 00 5b c3 66 90 <48> 8b 50 08 48 83 fa 01 76 dd 66 83 78 20 00 79 d6 f6 c2 01 74 18
--
2024-01-19T14:39:19.104754+01:00 slb-epg21 kernel: [  153.217213] haproxy[721]: segfault at 8 ip 000055c5fb9dd880 sp 00007f1e4adf32a0 error 4 in haproxy[55c5fb81c000+267000] likely on CPU 1 (core 0, socket 2)
2024-01-19T14:39:19.104758+01:00 slb-epg21 kernel: [  153.217234] Code: 00 00 48 8b 40 0f a8 01 74 20 66 83 78 1f 00 78 f1 48 85 c9 74 47 83 69 68 01 e8 cb 39 0a 00 c7 43 3c 00 00 00 00 5b c3 66 90 <48> 8b 50 08 48 83 fa 01 76 dd 66 83 78 20 00 79 d6 f6 c2 01 74 18
--
2024-01-19T14:49:00.100042+01:00 slb-epg21 kernel: [  734.160428] haproxy[937]: segfault at 8 ip 000055cad812c880 sp 00007ff0cb9f32a0 error 4 in haproxy[55cad7f6b000+267000] likely on CPU 1 (core 0, socket 2)
2024-01-19T14:49:00.100051+01:00 slb-epg21 kernel: [  734.160442] Code: 00 00 48 8b 40 0f a8 01 74 20 66 83 78 1f 00 78 f1 48 85 c9 74 47 83 69 68 01 e8 cb 39 0a 00 c7 43 3c 00 00 00 00 5b c3 66 90 <48> 8b 50 08 48 83 fa 01 76 dd 66 83 78 20 00 79 d6 f6 c2 01 74 18
--
2024-01-19T14:59:27.080018+01:00 slb-epg21 kernel: [ 1361.081982] haproxy[1003]: segfault at 8 ip 0000557bf4bf4880 sp 00007f9c1bdf32a0 error 4 in haproxy[557bf4a33000+267000] likely on CPU 1 (core 0, socket 2)
2024-01-19T14:59:27.080030+01:00 slb-epg21 kernel: [ 1361.082002] Code: 00 00 48 8b 40 0f a8 01 74 20 66 83 78 1f 00 78 f1 48 85 c9 74 47 83 69 68 01 e8 cb 39 0a 00 c7 43 3c 00 00 00 00 5b c3 66 90 <48> 8b 50 08 48 83 fa 01 76 dd 66 83 78 20 00 79 d6 f6 c2 01 74 18
--
2024-01-19T15:16:00.112094+01:00 slb-epg21 kernel: [ 2354.020706] haproxy[1060]: segfault at 8 ip 0000562bf44f0880 sp 00007fcfd71f32a0 error 4 in haproxy[562bf432f000+267000] likely on CPU 1 (core 0, socket 2)
2024-01-19T15:16:00.112103+01:00 slb-epg21 kernel: [ 2354.020729] Code: 00 00 48 8b 40 0f a8 01 74 20 66 83 78 1f 00 78 f1 48 85 c9 74 47 83 69 68 01 e8 cb 39 0a 00 c7 43 3c 00 00 00 00 5b c3 66 90 <48> 8b 50 08 48 83 fa 01 76 dd 66 83 78 20 00 79 d6 f6 c2 01 74 18


# /usr/sbin/haproxy -Ws -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -S /run/haproxy-master.sock
[NOTICE]   (1057) : haproxy version is 2.9.3-1~bpo12+1
[NOTICE]   (1057) : path to executable is /usr/sbin/haproxy
[NOTICE]   (1057) : New worker (1059) forked
[NOTICE]   (1057) : Loading success.
[NOTICE]   (1057) : haproxy version is 2.9.3-1~bpo12+1
[NOTICE]   (1057) : path to executable is /usr/sbin/haproxy
[ALERT]    (1057) : Current worker (1059) exited with code 139 (Segmentation fault)
[WARNING]  (1057) : A worker process unexpectedly died and this can only be explained by a bug in haproxy or its dependencies.
Please check that you are running an up to date and maintained version of haproxy and open a bug report.
HAProxy version 2.9.3-1~bpo12+1 2024/01/18 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2025.
Known bugs: http://www.haproxy.org/bugs/bugs-2.9.3.html
Running on: Linux 6.1.0-17-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.69-1 (2023-12-30) x86_64
[ALERT]    (1057) : exit-on-failure: killing every processes with SIGTERM
[WARNING]  (1057) : All workers exited. Exiting... (139)
```

**Comment by joachimbuechse** [comment]
I assume without a core dump it'll be tricky to get to the bottom of this, unfortunately I did not get any core dumps from those shut downs. I verified that core dumping in general works by doing a kill -11 on haproxy which produced a core file. Any advice in how I can provide helpful traces / cores / etc welcome. We can set up the same combination of haproxy + upstream server + load client in GCP if required.

**Comment by wtarreau** [comment]
Thanks for your report! With the use of the full debug logs and the executable from Vincent's repo, I managed to find the location of the crash, it's in the cache, in `delete_entry()` while calling `eb32_next_dup()`:
```
0000000000223830 <delete_entry>:
  223830:       53                      push   %rbx
  223831:       48 89 fb                mov    %rdi,%rbx
  223834:       48 83 c7 18             add    $0x18,%rdi
  223838:       8b 43 7c                mov    0x7c(%rbx),%eax
  22383b:       85 c0                   test   %eax,%eax
  22383d:       74 31                   je     223870 <delete_entry+0x40>
  22383f:       48 89 fa                mov    %rdi,%rdx
  223842:       31 c9                   xor    %ecx,%ecx
  223844:       0f 1f 40 00             nopl   0x0(%rax)
  223848:       48 8b 42 18             mov    0x18(%rdx),%rax
  22384c:       a8 01                   test   $0x1,%al
  22384e:       75 10                   jne    223860 <delete_entry+0x30>
  223850:       eb 2e                   jmp    223880 <delete_entry+0x50>
  223852:       66 0f 1f 44 00 00       nopw   0x0(%rax,%rax,1)
  223858:       48 8b 40 0f             mov    0xf(%rax),%rax
  22385c:       a8 01                   test   $0x1,%al
  22385e:       74 20                   je     223880 <delete_entry+0x50>
  223860:       66 83 78 1f 00          cmpw   $0x0,0x1f(%rax)
  223865:       78 f1                   js     223858 <delete_entry+0x28>
  223867:       48 85 c9                test   %rcx,%rcx
  22386a:       74 47                   je     2238b3 <delete_entry+0x83>
  22386c:       83 69 68 01             subl   $0x1,0x68(%rcx)
  223870:       e8 cb 39 0a 00          call   2c7240 <eb_delete>
  223875:       c7 43 3c 00 00 00 00    movl   $0x0,0x3c(%rbx)
  22387c:       5b                      pop    %rbx
  22387d:       c3                      ret    
  22387e:       66 90                   xchg   %ax,%ax
->223880:       48 8b 50 08             mov    0x8(%rax),%rdx
  223884:       48 83 fa 01             cmp    $0x1,%rdx
  223888:       76 dd                   jbe    223867 <delete_entry+0x37>
  22388a:       66 83 78 20 00          cmpw   $0x0,0x20(%rax)
  22388f:       79 d6                   jns    223867 <delete_entry+0x37>
  223891:       f6 c2 01                test   $0x1,%dl
  223894:       74 18                   je     2238ae <delete_entry+0x7e>
  223896:       66 2e 0f 1f 84 00 00    cs nopw 0x0(%rax,%rax,1)
```

However at this point I have no idea why, but it seems directly related to the use of secondary entries in the cache. I think it would help the troubleshooting if you could comment out the `max-secondary-entries` and `process-vary` lines to confirm it doesn't crash anymore.

**Comment by rlebreton** [comment]
I managed to reproduced the crash and it indeed seems to only happen when secondary entries are used in the cache. I don't have a fix yet but at least it's really quick to reproduce, I'm on it.

**Comment by joachimbuechse** [comment]
I can confirm that with max-secondary-entries and process-vary removed and "del-header Vary" enabled I see no crash either.

**Comment by rlebreton** [comment]
It took some time to fully understand why I only had this crash when 'process-vary' was enabled but I finally figured it out. The following patch should fix it.
[0001-BUG-MEDIUM-cache-Fix-crash-when-deleting-secondary-e.patch.txt](https://github.com/haproxy/haproxy/files/14041362/0001-BUG-MEDIUM-cache-Fix-crash-when-deleting-secondary-e.patch.txt)

**Comment by wtarreau** [comment]
Now merged into 3.0, thanks Rémi!

**Comment by wtarreau** [comment]
2.9 was released and was the only one affected, now closing.
